File: day12/main.py
import re

def answer(**regs):
    def read(reg):
        if reg in regs:
            return regs[reg]
        return int(reg)

    ins = [line.split() for line in open('input.txt')]
    pc = 0
    while pc < len(ins):
        i = ins[pc]
        if i[0] == 'cpy':
            regs[i[2]] = read(i[1])
        elif i[0] == 'inc':
            regs[i[1]] = regs[i[1]] + 1
        elif i[0] == 'dec':
            regs[i[1]] = regs[i[1]] - 1
        elif i[0] == 'jnz':
            if read(i[1]):
                pc += int(i[2])
                continue
        pc += 1
            
    return regs['a']

print('Answer #1:', answer(a=0, b=0, c=0, d=0))
print('Answer #2:', answer(a=0, b=0, c=1, d=0))

File: day12/main.nim
import macros
import strutils

proc compile(input: string, setup: string): NimNode {.compiletime.} =
  var caseBody = newNimNode(nnkCaseStmt)
  caseBody.add parseExpr("pc")
  var pc = 0

  template addCase(text): typed =
    var branch = newNimNode(nnkOfBranch)
    branch.add parseExpr($pc)
    branch.add parseExpr(text)
    caseBody.add branch

  for line in input.split("\n"):
    var i = line.split(" ")
    case i[0]
    of "cpy":
      addCase(i[2] & " = " & i[1])
    of "inc":
      addCase("inc " & i[1])
    of "dec":
      addCase("dec " & i[1])
    of "jnz":
      let n = i[2].parseInt
      addCase("if " & i[1] & " > 0: pc += " & $n & "; continue")
    else:
      discard
    inc pc

  let e = newNimNode(nnkElse)
  e.add parseStmt("discard")
  caseBody.add e

  var loopBody = newStmtList()
  loopBody.add caseBody
  loopBody.add parseExpr("inc pc")

  var stmts = newStmtList()

  template addStmt(text): typed =
    stmts.add parseStmt(text)

  addStmt(setup)
  addStmt("var pc: int")
  var loop = newNimNode(nnkWhileStmt)
  loop.add parseExpr("pc < " & $pc)
  loop.add loopBody
  stmts.add loop

  addStmt "echo a"
  # echo stmts.repr
  result = stmts

macro compileFile(filename: string, setup: string): typed =
  compile(staticRead(filename.strval), setup.strval)

proc answer1 = compileFile("input.txt", "var a, b, c, d: int")
proc answer2 = compileFile("input.txt", "var a, b, c, d: int\nc = 1")
answer1()
answer2()
